### PR TITLE
Add materialized view support for BigQuery and Snowflake

### LINKED ISSUE #1190: Support for BigQuery materialized views
Similar to [feature request for snowflake](https://github.com/dataform-co/dataform/issues/1028),  add support for [BigQuery materialized views](https://cloud.google.com/bigquery/docs/materialized-views-intro)

### PR DESCRIPTION
**Pull request by canberkkoparal**
fixes #1190 and #710

### CODE REVIEW

File: core/table.ts [hunk 1]
```diff
@@ -285,6 +285,17 @@ export interface ITableConfig
    * If configured, records with matching unique key(s) will be updated, rather than new rows being inserted.
    */
   uniqueKey?: string[];
+
+  /**
+   * Only allowed when the table type is `view`.
+   * Only allowed when using Snowflake or BigQuery warehouse.
```
Comment (BenBirt) [inline core/table.ts:291]: nit: "Only valid when using Snowflake or BigQuery."

File: core/table.ts [hunk 2]
```diff
@@ -285,6 +285,17 @@ export interface ITableConfig
    * If configured, records with matching unique key(s) will be updated, rather than new rows being inserted.
    */
   uniqueKey?: string[];
+
+  /**
+   * Only allowed when the table type is `view`.
```
Comment (BenBirt) [inline core/table.ts:290]: nit: s/allowed/valid/

File: examples/common_v2/definitions/example_materialized_view.sqlx [hunk 3]
```diff
@@ -0,0 +1,3 @@
+config {type: "view", materialized: true}
```
Comment (BenBirt) [inline examples/common_v2/definitions/example_materialized_view.sqlx:1]: supernit: git complaining that this and other new files don't have trailing newlines

**Comment by BenBirt** [comment]
This looks great, thanks!

I know you won't be able to see them, but the failing tests are:

```
@dataform/integration/bigquery > run > project e2e                       FAILED

    AssertionError: expected 19 to deeply equal 18
        at Test.fn (tests/integration/bigquery.spec.ts:44:45)
        at runMicrotasks (<anonymous>)
        at processTicksAndRejections (internal/process/task_queues.js:93:5)
        at testing/test.ts:74:15
        at Test.run (testing/test.ts:63:9)
        at runTestOrSuite (testing/suite.ts:80:7)
        at async Promise.all (index 0)
        at Suite.run (testing/suite.ts:89:9)
        at runTestOrSuite (testing/suite.ts:80:7)
        at async Promise.all (index 0)


    Expected:

    18

    Actual:

    19

    Overall diff (expected -, actual +):

    18+   19
```

```
@dataform/integration/snowflake > run                                    FAILED

    AssertionError: expected 21 to deeply equal 20
        at Test.fn (tests/integration/snowflake.spec.ts:47:43)
        at runMicrotasks (<anonymous>)
        at processTicksAndRejections (internal/process/task_queues.js:93:5)
        at testing/test.ts:74:15
        at Test.run (testing/test.ts:63:9)
        at runTestOrSuite (testing/suite.ts:80:7)
        at Suite.run (testing/suite.ts:92:11)
        at async Promise.all (index 0)
        at Function.run (testing/runner.ts:57:7)


    Expected:

    20

    Actual:

    21

    Overall diff (expected -, actual +):

    20+   21
```

**Review by BenBirt — CHANGES_REQUESTED** [review]

**Comment by astorrs** [comment]
@canberkkoparal This is amazing! 

For BigQuery it would be good to expose setting the `enableRefresh` and `refreshIntervalMs` options to allow for finer control (https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#materializedviewdefinition). 

Otherwise an ALTER MATERIALIZED VIEW statement would have to be done in a post_operation in Dataform to adjust https://cloud.google.com/bigquery/docs/reference/standard-sql/data-definition-language#materialized_view_set_options_list). 

Any opinion on extending the config @BenBirt @lewish ?

As far as I’m aware Snowflake doesn’t currently allow you to tune the refresh rate in the same way.

**Comment by BenBirt** [comment]
@astorrs I'm all up for extending configuration options further, but: it seems that the default for `enableRefresh` is `true`; why would a user ever want to disable it (for Dataform-managed MVs, at least)? (this just sounds like `type: "table"` to me)

Allowing configurability of `refreshIntervalMs` makes sense. @canberkkoparal WDYT?

**Comment by astorrs** [comment]
@BenBirt you're probably right on the `enableRefresh`; the only time I’ve used it is when you want to trigger a refresh based on some external event and with Dataform driving it’s redundant. 

`refreshIntervalMs` is more important since we run our Dataform jobs more frequently than every 30 mins (which is the default max refresh window) and I want them to be at least that fresh. The unique benefit of materialized views in BQ is the way it seamlessly incorporates data from the streaming buffer into the view in between refreshes.

**Comment by BenBirt** [comment]
@astorrs just to be clear I'm assuming you mean at *most* that fresh?

**Comment by BenBirt** [comment]
@astorrs took me a couple of reads but that makes sense, thanks!

Given that BQ DDL only lets us specify this interval in minutes, we'll probably go with something more like `refreshIntervalMins`, but SGTM.

**Review by BenBirt — APPROVED** [review]
Update https://github.com/dataform-co/dataform/blob/master/version.bzl to `1.21.0`? Then I'll publish a new release.

**Pull request merged**